src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CompositeValueProvider.cs

```diff
@@ -184,5 +205,63 @@ public IValueProvider Filter(BindingSource bindingSource)
 
             return new CompositeValueProvider(filteredValueProviders);
         }
+
+        /// <inheritdoc />
+        /// <remarks>
+        /// Value providers are included by default. If a contained <see cref="IValueProvider"/> does not implement
+        /// <see cref="IKeyRewriterValueProvider"/>, <see cref="Filter()"/> will not remove it.
+        /// </remarks>
+        public IValueProvider Filter()
+        {
+            var shouldFilter = false;
+            for (var i = 0; i < Count; i++)
+            {
+                var valueProvider = Items[i];
+                if (valueProvider is IKeyRewriterValueProvider)
+                {
+                    shouldFilter = true;
+                    break;
+                }
+            }
+
+            if (!shouldFilter)
+            {
+                // No inner IKeyRewriterValueProvider implementations. Nothing to exclude.
+                return this;
+            }
+
+            var filteredValueProviders = new List<IValueProvider>();
+            for (var i = 0; i < Count; i++)
+            {
+                var valueProvider = Items[i];
+                if (valueProvider is IKeyRewriterValueProvider keyRewriterValueProvider)
+                {
+                    var result = keyRewriterValueProvider.Filter();
+                    if (result != null)
+                    {
+                        filteredValueProviders.Add(result);
+                    }
+                }
+                else
+                {
+                    // Assume value providers that aren't rewriter-aware do not rewrite their keys.
+                    filteredValueProviders.Add(valueProvider);
+                }
+            }
+
+            if (filteredValueProviders.Count == 0)
+            {
+                // Do not create an empty CompositeValueProvider.
+                return null;
+            }
+
+            if (filteredValueProviders.Count == Count)
+            {
+                // No need for a new CompositeValueProvider because no rewriter-aware provider excluded itself.
+                return this;
+            }
+
+            return new CompositeValueProvider(filteredValueProviders);
```

Comment:
Do we really need all of the extra complexity here compared to just coding this up in a straightforward way? Worst case this is a single extra list allocation per request?

Reply:
The `IBindingSourceValueProvider` code was going to some lengths to avoid allocating a `CompositeValueProvider` but ignoring the larger list allocation. I just updated the existing method to be less alloc-y and wrote about the same code in the new method. If we don't think allocating a list per viewed collection property is important, fine. Are you suggesting `return this` is also a premature optimization in both methods?